File: 0656_coin_path/solution.py
# LeetCode 0656 - Coin Path
# https://leetcode.com/problems/coin-path/

from typing import List


class Solution:
    def cheapestJump(self, coins: List[int], maxJump: int) -> List[int]:
        n = len(coins)
        if coins[-1] == -1:
            return []

        inf = 10**18
        cost = [inf] * n
        nxt = [-1] * n
        cost[-1] = coins[-1]

        for i in range(n - 2, -1, -1):
            if coins[i] == -1:
                continue
            for jump in range(1, maxJump + 1):
                j = i + jump
                if j >= n:
                    break
                if cost[j] == inf:
                    continue
                candidate = coins[i] + cost[j]
                if candidate < cost[i] or (
                    candidate == cost[i] and (nxt[i] == -1 or j < nxt[i])
                ):
                    cost[i] = candidate
                    nxt[i] = j

        if cost[0] == inf:
            return []

        path = [1]
        i = 0
        while i != n - 1:
            i = nxt[i]
            path.append(i + 1)
        return path

File: 0656_coin_path/Solution.scala
// LeetCode 0656 - Coin Path
// https://leetcode.com/problems/coin-path/

import scala.collection.mutable

object Solution {
  def cheapestJump(coins: Array[Int], maxJump: Int): List[Int] = {
    val n = coins.length
    if (coins(n - 1) == -1) return List.empty
    val inf = Long.MaxValue / 4
    val cost = Array.fill(n)(inf)
    val nxt = Array.fill(n)(-1)
    cost(n - 1) = coins(n - 1)
    var i = n - 2
    while (i >= 0) {
      if (coins(i) != -1) {
        var jump = 1
        while (jump <= maxJump) {
          val j = i + jump
          if (j >= n) jump = maxJump + 1
          else {
            if (cost(j) != inf) {
              val candidate = coins(i) + cost(j)
              if (candidate < cost(i) || (candidate == cost(i) && (nxt(i) == -1 || j < nxt(i)))) {
                cost(i) = candidate
                nxt(i) = j
              }
            }
            jump += 1
          }
        }
      }
      i -= 1
    }
    if (cost(0) == inf) return List.empty
    val path = mutable.ArrayBuffer(1)
    i = 0
    while (i != n - 1) {
      i = nxt(i)
      path += i + 1
    }
    path.toList
  }
}
